apps/meeting/src/providers/VideoGridProvider/state.tsx

```diff
@@ -0,0 +1,510 @@
+}
+interface VideoSourceState {
+  cameraSources: string[];
+  activeSpeakersWithCameraSource: string[];
+  contentShareId: string | null;
+  hasLocalVideo: boolean;
+  hasLocalContentSharing: boolean;
+}
+
+type VideoSourceWithType = { attendeeId: string; type: string };
+
+export type State = {
+  gridState: GridState;
+  attendeeStates: { [attendeeId: string]: AttendeeState };
+  videoSourceState: VideoSourceState;
+};
+
+export const initialState: State = {
+  gridState: {
+    mode: VideoGridMode.GalleryView,
+    isZoomed: false,
+    threshold: 8,
+  },
+  attendeeStates: {},
+  videoSourceState: {
+    cameraSources: [],
+    activeSpeakersWithCameraSource: [],
+    contentShareId: null,
+    hasLocalVideo: false,
+    hasLocalContentSharing: false,
+  },
+};
+
+export type RosterAttendeeType = {
+  chimeAttendeeId: string;
+  externalUserId?: string;
+  name?: string;
+};
+
+export type RosterType = {
+  [attendeeId: string]: RosterAttendeeType;
+};
+
+export type Controls = {
+  zoomIn: () => void;
+  zoomOut: () => void;
+};
+
+export enum VideoGridAction {
+  ResetVideoGridState,
+  UpdateAttendeeStates,
+  UpdateVideoSources,
+  UpdateActiveSpeakers,
+  UpdateLocalSourceState,
+  UpdateGridState,
+  PauseVideoTile,
+  UnpauseVideoTile,
+  ZoomIn,
+  ZoomOut,
+}
+
+type ResetVideoGridState = {
+  type: VideoGridAction.ResetVideoGridState;
+  payload?: any;
+};
+
+type UpdateVideoSources = {
+  type: VideoGridAction.UpdateVideoSources;
+  payload: {
+    videoSources: VideoSource[];
+  };
+};
+
+type UpdateActiveSpeakers = {
+  type: VideoGridAction.UpdateActiveSpeakers;
+  payload: {
+    activeSpeakers: string[];
+  };
+};
+
+type UpdateGridState = {
+  type: VideoGridAction.UpdateGridState;
+  payload: {
+    videoGridMode: VideoGridMode;
+  };
+};
+
+type UpdateAttendeeStates = {
+  type: VideoGridAction.UpdateAttendeeStates;
+  payload: {
+    roster: RosterType;
+  };
+};
+
+type UpdateLocalSourceState = {
+  type: VideoGridAction.UpdateLocalSourceState;
+  payload: {
+    isVideoEnabled: boolean;
+    localAttendeeId: string | null;
+    isLocalUserSharing: boolean;
+    sharingAttendeeId: string | null;
+  };
+};
+
+type PauseVideoTile = {
+  type: VideoGridAction.PauseVideoTile;
+  payload: {
+    attendeeId: string;
+  };
+};
+
+type UnpauseVideoTile = {
+  type: VideoGridAction.UnpauseVideoTile;
+  payload: {
+    attendeeId: string;
+  };
+};
+
+type ZoomIn = {
+  type: VideoGridAction.ZoomIn;
+  payload?: any;
+};
+
+type ZoomOut = {
+  type: VideoGridAction.ZoomOut;
+  payload?: any;
+};
+
+export type Action =
+  | UpdateVideoSources
+  | ResetVideoGridState
+  | UpdateGridState
+  | UpdateAttendeeStates
+  | UpdateLocalSourceState
+  | UpdateActiveSpeakers
+  | PauseVideoTile
+  | UnpauseVideoTile
+  | ZoomIn
+  | ZoomOut;
+
+const isContentShare = (sourceId: string): boolean =>
+  new DefaultModality(sourceId).hasModality(DefaultModality.MODALITY_CONTENT);
+
+const calculateVideoSourcesToBeRendered = (
```

Comment:
Should we make a new `Utils.ts` files in `VideoGridProvider/Utils.ts` and move `updateDownlinkPreferences`, `calculateVideoSourcesToBeRendered` and `isContentShare` there?